Initial support for loongarch64

qiangxuhui opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Experimentally support loongarch64 as a new platform for ClickHouse.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

qiangxuhui committed:
Initial support for loongarch64

Make ClickHouse compilable and runnable on loongarch64

So far only basic functionality was tested (on real hw),
clickhouse server runs, exceptions works, client works,
simple tests works.

qiangxuhui commented:
## Some things that need help

### About `contrib/double-conversion`

The [google/double-conversion](https://github.com/google/double-conversion) support loongarch64 at [loongarch-supported-commit](https://github.com/google/double-conversion/commit/32bc443c60c860eb6b4843533a614766d611172e). But the [commit](https://github.com/google/double-conversion/tree/cf2f0f3d547dc73b4612028a155b80536902ba02) used by `contrib/double-conversion` lags behind [loongarch-supported-commit](https://github.com/google/double-conversion/commit/32bc443c60c860eb6b4843533a614766d611172e) too much. If [loongarch-supported-commit](https://github.com/google/double-conversion/commit/32bc443c60c860eb6b4843533a614766d611172e) is used, it will cause many build errors.

So i think a appropriate way to solve this problem is to fork [contrib/double-conversion](https://github.com/google/double-conversion) to ClickHouse organization, like [contrib/boringssl](https://github.com/ClickHouse/boringssl).

> The URL currently used by `contrib/double-conversion` in this patch is https://github.com/qiangxuhui/boringssl/commit/733eab0190a71b734be99c1a57efb4d1bf997086

### About `src/Common/ThreadFuzzer.cpp`

```sh
## linux-x86_64
root@clickhouse-cross /w/c/ClickHouse (loongarch64)# readelf -sW  contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc/lib64/libpthread.so.0 | grep "pthread_mutex_lock"
   220: 000000000000bfc0   636 FUNC    WEAK   DEFAULT   17 pthread_mutex_lock@@GLIBC_2.2.5
   247: 000000000000bfc0   636 FUNC    GLOBAL DEFAULT   17 __pthread_mutex_lock@@GLIBC_2.2.5
    40: 000000000000b8a0  1821 FUNC    LOCAL  DEFAULT   17 __pthread_mutex_lock_full
   193: 000000000000bfc0   636 FUNC    LOCAL  DEFAULT   17 __GI___pthread_mutex_lock
   408: 000000000000bfc0   636 FUNC    WEAK   DEFAULT   17 pthread_mutex_lock
   542: 000000000000bfc0   636 FUNC    GLOBAL DEFAULT   17 __pthread_mutex_lock

## linux-loongarch64
root@clickhouse-cross /w/c/ClickHouse (loongarch64)# readelf -sW contrib/sysroot/linux-loongarch64/lib/loongarch64-linux-gnu/libc.so.6 | grep "pthread_mutex_lock"
  1265: 0000000000081578   828 FUNC    GLOBAL DEFAULT   13 pthread_mutex_lock@@GLIBC_2.36
```

As seen above, the `pthread_mutex_lock` function in the loongarch64 platform does not have an internal symbol similar to "__pthread_mutex_lock." Therefore, I made the following modification:

```diff
-#if defined(OS_LINUX) && !defined(THREAD_SANITIZER) && !defined(MEMORY_SANITIZER)
+#if defined(OS_LINUX) && !defined(THREAD_SANITIZER) && !defined(MEMORY_SANITIZER)  && !defined(__loongarch64)
     #define THREAD_FUZZER_WRAP_PTHREAD 1
 #else
     #define THREAD_FUZZER_WRAP_PTHREAD 0
```

I am not sure whether this modification is appropriate and hope to get some advice.

qiangxuhui commented:
```sh
root@fedora-clickhouse /t/work# uname -m
loongarch64
root@fedora-clickhouse /t/work# file clickhosue-cross
clickhosue-cross: ELF 64-bit LSB executable, LoongArch, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-loongarch-lp64d.so.1, for GNU/Linux 5.19.0, BuildID[sha1]=40222f7d7b66ec56fdb1be2fc76aa654c30e6b5a, with debug_info, not stripped


root@fedora-clickhouse /t/work# ./clickhosue-cross server
Processing configuration file 'config.xml'.
There is no file 'config.xml', will use embedded config.
2024.03.14 17:21:03.336279 [ 1074218 ] {} <Information> SentryWriter: Sending crash reports is disabled
2024.03.14 17:21:03.336376 [ 1074218 ] {} <Trace> Pipe: Pipe capacity is 1.00 MiB
...
2024.03.14 17:21:04.143717 [ 1074218 ] {} <Information> Application: Listening for http://[::1]:8123
2024.03.14 17:21:04.143809 [ 1074218 ] {} <Information> Application: Listening for native protocol (tcp): [::1]:9000
2024.03.14 17:21:04.143887 [ 1074218 ] {} <Information> Application: Listening for MySQL compatibility protocol: [::1]:9004
2024.03.14 17:21:04.143964 [ 1074218 ] {} <Information> Application: Listening for http://127.0.0.1:8123
2024.03.14 17:21:04.144057 [ 1074218 ] {} <Information> Application: Listening for native protocol (tcp): 127.0.0.1:9000
2024.03.14 17:21:04.144135 [ 1074218 ] {} <Information> Application: Listening for MySQL compatibility protocol: 127.0.0.1:9004
2024.03.14 17:21:04.144167 [ 1074218 ] {} <Information> Application: Ready for connections.
```

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #61509 from qiangxuhui/ClickHouse:loongarch64 -->
*This is an automated comment for commit e8a781032c2ea294146b35df04c46836c718a69e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61509/e8a781032c2ea294146b35df04c46836c718a69e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

qiangxuhui commented:
> You can add cross compiled build to CI as well. As we build for darwin

The toolchain used by Loongarch64 is based on Linux, so I added CI build support with reference to riscv64.

qiangxuhui commented:
@alexey-milovidov PTAL.

alexey-milovidov commented:
@qiangxuhui, I've forked double-conversion to https://github.com/ClickHouse/double-conversion
Please send a PR there.

alexey-milovidov commented on src/Common/ThreadFuzzer.cpp:
```diff
@@ -22,7 +22,7 @@
 
 /// We will also wrap some thread synchronization functions to inject sleep/migration before or after.
-#if defined(OS_LINUX) && !defined(THREAD_SANITIZER) && !defined(MEMORY_SANITIZER)
+#if defined(OS_LINUX) && !defined(THREAD_SANITIZER) && !defined(MEMORY_SANITIZER)  && !defined(__loongarch64)
```
Why not?

qiangxuhui replied:
The `pthread_mutex_lock` function in the loongarch64 platform does not have an internal symbol similar to "__pthread_mutex_lock@{GLIBC_VERSION}." The loongarch64's format is:
```sh
root@clickhouse-cross /w/c/ClickHouse (loongarch64)# readelf -sW contrib/sysroot/linux-loongarch64/lib/loongarch64-linux-gnu/libc.so.6 | grep "pthread_mutex_lock"
  1265: 0000000000081578   828 FUNC    GLOBAL DEFAULT   13 pthread_mutex_lock@@GLIBC_2.36
```
This is different from x86-64, arm64, riscv64 etc.

qiangxuhui replied:
This modification is because I thought before that ThreadFuzzer could not intercept the "pthrad_mutex_lock" function without the "__pthrad_mutex_lock" symbol.

After referring to https://github.com/ClickHouse/ClickHouse/commit/bdeb5514c5e49b2f6158ca5b1f519956e55a03ec, I realized that ThreadFuzzer can intercept `pthrad_mutex_lock` through `MAKE_WRAPPER_USING_DLSYM`. So I will modify my code.

alexey-milovidov reviewed (commented):
Everything else looks good, and we can launch the CI soon.

qiangxuhui commented:
> @qiangxuhui, I've forked double-conversion to https://github.com/ClickHouse/double-conversion Please send a PR there.

@alexey-milovidov， it is necessary to create a branch used by clickhouse based on https://github.com/ClickHouse/double-conversion/commit/cf2f0f3d547dc73b4612028a155b80536902ba02

alexey-milovidov commented:
https://github.com/ClickHouse/double-conversion/tree/clickhouse

qiangxuhui commented:
@alexey-milovidov I have modified `src/Common/ThreadFuzzer.cpp`, please Take A Look.

rschu1ze commented on src/Disks/IO/ThreadPoolReader.cpp:
```diff
@@ -36,6 +36,8 @@
         #define SYS_preadv2 380
     #elif defined(__riscv)
         #define SYS_preadv2 286
+    #elif defined(__loongarc64)
```
`__loongarc64`? Should that not be `__loongarch64`?

qiangxuhui replied:
That should be `loongarch64`, it's my mistake.

rschu1ze commented on .github/workflows/master.yml:
```diff
@@ -242,6 +242,14 @@ jobs:
       build_name: binary_s390x
       data: ${{ needs.RunConfig.outputs.data }}
       checkout_depth: 0
+  BuilderBinLOONGARCH64:
```
Unrelated to this file:

To try the loongarch port, I ran locally

```
$ rm -rf build && cmake -DCMAKE_TOOLCHAIN_FILE=cmake/linux/toolchain-loongarch64.cmake -DENABLE_EMBEDDED_COMPILER=1 -DENABLE_LIBRARIES=1 -DCMAKE_BUILD_TYPE=Debug -DPARALLEL_COMPILE_JOBS=64 -DPARALLEL_LINK_JOBS=64 -S . -B build
```

which is the standard CMake invocation for cross-compiling. This was the output:

```
~/repo/ch6 (loongarch64 =) $ rm -rf build && cmake -DCMAKE_TOOLCHAIN_FILE=cmake/linux/toolchain-loongarch64.cmake -DENABLE_EMBEDDED_COMPILER=1 -DENABLE_LIBRARIES=1 -DCMAKE_BUILD_TYPE=Debug -DPARALLEL_COMPILE_JOB
S=64 -DPARALLEL_LINK_JOBS=64 -S . -B build
-- The C compiler identification is Clang 17.0.6
-- The CXX compiler identification is Clang 17.0.6
-- The ASM compiler identification is Clang with GNU-like command-line
-- Found assembler: /usr/bin/clang-17
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: /usr/bin/clang-17
-- Check for working C compiler: /usr/bin/clang-17 - broken
CMake Error at /usr/share/cmake-3.22/Modules/CMakeTestCCompiler.cmake:69 (message):
  The C compiler

    "/usr/bin/clang-17"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/ubuntu/repo/ch6/build/CMakeFiles/CMakeTmp

    Run Build Command(s):/usr/bin/ninja cmTC_d8b4c && [1/2] Building C object CMakeFiles/cmTC_d8b4c.dir/testCCompiler.c.o
    FAILED: CMakeFiles/cmTC_d8b4c.dir/testCCompiler.c.o
    /usr/bin/clang-17 --target=loongarch64-linux-gnu --sysroot=/home/ubuntu/repo/ch6/cmake/linux/../../contrib/sysroot/linux-loongarch64   -mcmodel=extreme --gcc-toolchain=/home/ubuntu/repo/ch6/cmake/linux/../..
/contrib/sysroot/linux-loongarch64/usr -MD -MT CMakeFiles/cmTC_d8b4c.dir/testCCompiler.c.o -MF CMakeFiles/cmTC_d8b4c.dir/testCCompiler.c.o.d -o CMakeFiles/cmTC_d8b4c.dir/testCCompiler.c.o -c /home/ubuntu/repo/ch
6/build/CMakeFiles/CMakeTmp/testCCompiler.c
    clang-17: error: invalid argument 'extreme' to -mcmodel=
    ninja: build stopped: subcommand failed.





  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:3 (project)
```

rschu1ze commented on .gitmodules:
```diff
@@ -23,8 +23,9 @@
+	branch = "ch-loongarch"
 	path = contrib/double-conversion
-	url = https://github.com/google/double-conversion
+	url = https://github.com/qiangxuhui/double-conversion
```
Can we instead please create a fork of double-conversion in ClickHouse's GitHub organization? (https://github.com/ClickHouse/). At last that would be the standard practice.

qiangxuhui replied:
I have sended a PR， https://github.com/ClickHouse/double-conversion/pull/1/files.

rschu1ze replied:
Thanks. Please see my comment [here](https://github.com/ClickHouse/double-conversion/pull/1). Additionally, we'd need to switch to `url = https://github.com/ClickHouse/double-conversion`. This can be done by (1) `git rm contrib/double-conversion` + committing the result, followed by (2) `cd contrib/`, `git submodule add --force https://github.com/ClickHouse/double-conversion.git` + committing the result, followed by (3) `cd double-conversion`, `git switch --create clickhouse origin/clickhouse`, `cd ..`, followed by a commit.

EDIT: This seems to get complicated. I'd be good with keeping `url = https://github.com/qiangxuhui/double-conversion` for now. When I am back from vacation, I can fix this up.

rschu1ze replied:
Fixed: https://github.com/ClickHouse/ClickHouse/pull/62177

rschu1ze commented on base/poco/Foundation/include/Poco/Platform.h:
```diff
@@ -120,6 +120,8 @@
 #define POCO_ARCH_AARCH64 0x0f
 #define POCO_ARCH_ARM64 0x0f // same as POCO_ARCH_AARCH64
 #define POCO_ARCH_RISCV64 0x10
+#define POCO_ARCH_RISCV32 0x11
```
Wondering why constants for RISC-V 32 are defined? That sounds fairly unrelated.

qiangxuhui replied:
I wanted to be consistent with [poco's upstream](https://github.com/pocoproject/poco/blob/devel/Foundation/include/Poco/Platform.h#L138), which doesn't seem necessary right now.

rschu1ze commented on base/poco/Foundation/include/Poco/Platform.h:
```diff
@@ -206,8 +208,16 @@
+	#if (__riscv_xlen == 64)
+		#define POCO_ARCH POCO_ARCH_RISCV64
+		#define POCO_ARCH_LITTLE_ENDIAN 1
+	#elif(__riscv_xlen == 32)
```
Same here, why are defines for RISC-V 32bit defined? Do we ever intend to support it?

qiangxuhui replied:
I wanted to be consistent with [poco's upstream](https://github.com/pocoproject/poco/blob/devel/Foundation/include/Poco/Platform.h#L138), which doesn't seem necessary right now.

qiangxuhui replied:
Same reason as above.

rschu1ze commented on cmake/linux/toolchain-loongarch64.cmake:
```diff
@@ -0,0 +1,21 @@
+set (CMAKE_CXX_COMPILER_TARGET "loongarch64-linux-gnu")
+set (CMAKE_ASM_COMPILER_TARGET "loongarch64-linux-gnu")
+
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -mcmodel=extreme")
```
Please add a comment, otherwise future readers will have no idea why `-mcmodel=extreme` is needed for loongarch64.

qiangxuhui replied:
OK, I'll add a comment explaining why -mcmodel=extreme is needed for loongarch64.

rschu1ze commented on contrib/librdkafka-cmake/config.h.in:
```diff
@@ -66,7 +66,7 @@
 #cmakedefine WITH_SASL_CYRUS 1
 // crc32chw
-#if !defined(__PPC__) && !defined(__riscv) && !defined(__aarch64__) && !defined(__s390x__)
+#if !defined(__PPC__) && !defined(__riscv) && !defined(__aarch64__) && !defined(__s390x__) && !defined(__loongarch64)
```
Related to the changes in contrib/boringssl/.

Please don't submit this until https://github.com/ClickHouse/ClickHouse/pull/59870 is merged. The other PR takes precedence for now.

rschu1ze commented on docs/en/development/build-cross-loongarch.md:
````diff
@@ -0,0 +1,30 @@
+``` bash
+cd ClickHouse
+mkdir build-loongarch64
+CC=clang-18 CXX=clang++-18 cmake . -Bbuild-loongarch64 -G Ninja -DCMAKE_TOOLCHAIN_FILE=cmake/linux/toolchain-loongarch64.cmake
````
Oh, I guess I need clang-18 to make loongarch64 compile locally (see my above comment). Maybe add a comment that clang-17 is not supported.

qiangxuhui replied:
I agree with your suggestion.

rschu1ze commented:
(@qiangxuhui Side note: Kindly don't *force*-push after reviews have been done, instead please add fixup commits on top and push regularly. Otherwise, there is no way for reviewers to track the change history of the PR - thanks)

qiangxuhui commented:
> (@qiangxuhui Side note: Kindly don't _force_-push after reviews have been done, instead please add fixup commits on top and push regularly. Otherwise, there is no way for reviewers to track the change history of the PR - thanks)

@rschu1ze , thanks, I will follow your suggestions in future submissions, is there anything else I need to do at the moment?

rschu1ze commented:
> @rschu1ze , thanks, I will follow your suggestions in future submissions, is there anything else I need to do at the moment?

Currently, only bugfixes are allowed in the codebase, so we'd have to wait with merging a bit more.

Also, https://github.com/ClickHouse/ClickHouse/pull/59870 and https://github.com/ClickHouse/ClickHouse/pull/62177 are prerequisites to merging this PR.

About double-conversion: Please make another PR similar to https://github.com/ClickHouse/double-conversion/pull/1/files but against branch `ClickHouse/v3.3.0` in [this repository](https://github.com/ClickHouse/double-conversion/tree/ClickHouse/v3.3.0). After https://github.com/ClickHouse/ClickHouse/pull/62177, `.gitmodules` will point to https://github.com/ClickHouse/double-conversion, let's keep it that way.

All your other changes look good - thanks.

qiangxuhui commented:
> Currently, only bugfixes are allowed in the codebase, so we'd have to wait with merging a bit more.
> 
> Also, #59870 and #62177 are prerequisites to merging this PR.
> 
> About double-conversion: Please make another PR similar to https://github.com/ClickHouse/double-conversion/pull/1/files but against branch `ClickHouse/v3.3.0` in [this repository](https://github.com/ClickHouse/double-conversion/tree/ClickHouse/v3.3.0). After #62177, `.gitmodules` will point to https://github.com/ClickHouse/double-conversion, let's keep it that way.
> 
> All your other changes look good - thanks.

The branch [ClickHouse/v3.3.0](https://github.com/ClickHouse/double-conversion/tree/ClickHouse/v3.3.0) already contains [code support for loongarch](https://github.com/ClickHouse/double-conversion/blob/ClickHouse/v3.3.0/double-conversion/utils.h#L142). There is no need to make a PR again.

qiangxuhui committed:
Add support for building with openssl for loongarch64

rschu1ze commented on .github/workflows/master.yml:
```diff
@@ -247,6 +247,14 @@ jobs:
   #     data: ${{ needs.RunConfig.outputs.data }}
   #     checkout_depth: 0
+  BuilderBinLOONGARCH64:
+    needs: [RunConfig, BuilderDebRelease]
```
We need to have loongarch64 run in the CI infrastructure, otherwise the chance that the builds breaks is too big.

There are more changes required to enable such builds.
Please take a look at [this PR](https://github.com/ClickHouse/ClickHouse/pull/59870/commits/116393c9516d7ea33544dd06488f07bcd5ca7c9b) of mine (which disables s390x builds) to find the relevant places.

rschu1ze commented on contrib/openssl-cmake/CMakeLists.txt:
```diff
@@ -199,6 +202,13 @@ elseif(ARCH_RISCV64)
     perl_generate_asm(${OPENSSL_SOURCE_DIR}/crypto/modes/asm/ghash-riscv64.pl ${OPENSSL_BINARY_DIR}/crypto/modes/ghash-riscv64.S)
+elseif(ARCH_LOONGARCH64)
+    macro(perl_generate_asm FILE_IN FILE_OUT)
+        add_custom_command(OUTPUT ${FILE_OUT}
```
Please apply the same changes as done [here](https://github.com/ClickHouse/ClickHouse/pull/62754) to the invocation of `add_custom_command` - thanks.

qiangxuhui replied:
Ok.

rschu1ze reviewed (commented):
I wonder why the many changes in rocksdb were necessary? (they seem to go beyond loongarch64 support)

qiangxuhui commented on base/poco/Foundation/include/Poco/Platform.h:
```diff
@@ -209,6 +210,8 @@
 #    define POCO_ARCH POCO_ARCH_RISCV64
 #    define POCO_ARCH_LITTLE_ENDIAN 1
+#elif defined(__loongarch64)
+#    define POCO_ARCH POCO_ARCH_LOONGARCH64
```
Sorry, it was my mistake.

qiangxuhui committed:
Fix code for loongarch64

Fix `base/poco` and `contrib/openssl-cmake` based on code review comments(https://github.com/ClickHouse/ClickHouse/pull/61509).

qiangxuhui commented:
> I wonder why the many changes in rocksdb were necessary? (they seem to go beyond loongarch64 support)

Currently, this patch refers to the code of the [clickhouse/rocksdb master branch](https://github.com/ClickHouse/rocksdb). The redundant code changes are from "https://github.com/ClickHouse/rocksdb/commit/6efdb8d673ad2a4d9ac01cff22e2450fcf4948da" and "https://github.com/ClickHouse/rocksdb/commit/e83c37ff82ffeb2522bfb0f32dbd0af76d84a510 ". 

So I send a PR https://github.com/ClickHouse/rocksdb/pull/23 based on `clickhouse/rocksdb branch remove-regex-hdr`, after this PR is merged, I will change the reference to `clickhouse/rocksdb` and the redundant changes will be autoremoved.

rschu1ze commented:
@qiangxuhui This PR looks generally good to me (thanks!). Please make sure that there is a loongarch64 build in CI (please see my [previous comment](https://github.com/ClickHouse/ClickHouse/pull/61509#discussion_r1572423781) for this)

qiangxuhui committed:
Fix rocksdb's version

qiangxuhui commented:
> @qiangxuhui This PR looks generally good to me (thanks!). Please make sure that there is a loongarch64 build in CI (please see my [previous comment](https://github.com/ClickHouse/ClickHouse/pull/61509#discussion_r1572423781) for this)

@rschu1ze The minimum clang version required to build loongarch64 is clang-18, I noticed this PR https://github.com/ClickHouse/ClickHouse/pull/60469 is promoting the CI's clang version to clang-18, so I think it is more appropriate to add loongarch64 ci support after https://github.com/ClickHouse/ClickHouse/pull/60469 is merged. What do you think?

rschu1ze commented:
@qiangxuhui Yes, let's wait for clang-18 then.

rschu1ze commented:
The existing `loongarch64` [branch](https://github.com/ClickHouse/ClickHouse/tree/loongarch64) looked stale and I deleted it.

Anyways, let's continue with [your other PR](https://github.com/ClickHouse/ClickHouse/pull/63733) ...
